### PR TITLE
feat(ingest/powerbi): support modified_since, extract_dataset_schema and many more

### PR DESCRIPTION
**Pull request by aezomz**
## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

1. modified_since config ingestion which ingest only modified workspaces (which also improve stateful ingestion to be per checkpoint per workspace)
2. ownership config to have more fine grain control of how the corp user is built and also what criteria in term of authority over the powerbi asset
3. extract_datasets_to_containers config wrap powerbi tables (datahub dataset) under 1 powerbi dataset (datahub container)
4. extract_only_matched_endorsed_dataset config to extract only dataset that have matching endorsement
5. extract_dashboards config whether to ingest powerbi dashboards
6. extract_dataset_schema to extract powerbi dataset table schema to datahub dataset schema

I think the main improvement will be modified_since, which leave a interesting problem for stateful.
I tried my best to solve this in the most elegant way as compared to my previous PR which got rejected.

As much as I am catering for backward compatibility, those extra config are entirely optional and will not impact current users. :)

### CODE REVIEW

**Comment by aezomz** [comment]
@hsheth2 , @mohdsiddique and @jjoyce0510  . Can you all please help to take a look?

File: metadata-ingestion/src/datahub/ingestion/source/state/stale_entity_removal_handler.py
Comment (mohdsiddique) [inline metadata-ingestion/src/datahub/ingestion/source/state/stale_entity_removal_handler.py:196]: Could you please add comment for this code change

File: metadata-ingestion/tests/integration/powerbi/golden_test_admin_access_not_allowed.json
Comment (mohdsiddique) [inline metadata-ingestion/tests/integration/powerbi/golden_test_admin_access_not_allowed.json:9]: ownership should not break the old golden files. The  code should be backward compatible.

**Review by mohdsiddique — CHANGES_REQUESTED** [review]

File: metadata-ingestion/tests/integration/powerbi/test_m_parser.py
Comment (aezomz) [reply metadata-ingestion/tests/integration/powerbi/test_m_parser.py:198]: public_consumer_price_index is the table name

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py
Comment (aezomz) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:253]: This is good recommendation, i will create a datahubDataType attribute

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/powerbi_api.py
Comment (aezomz) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/powerbi_api.py:368]: didn't change anything actually. just indentation. cause of the for loop workspace

**Comment by aezomz** [comment]
@mohdsiddique  hihi, please help me to review again. I have changed accordingly to your comments. thanks!

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py
Comment (mohdsiddique) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:237]: We could do `isinstance` check here instead of attribute check. for example isinstance of Column then determine the description from expression and description fields

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py
Comment (mohdsiddique) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:232]: Can we rename this to to_datahub_schema_field

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py
Comment (mohdsiddique) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:273]: Please debug level log message, like `Returning empty dataset_mcps as no dataset tag matched with extract_only_matched_endorsed_dataset"

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py
Comment (mohdsiddique) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:292]: Add "expression" in constant

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py
Comment (mohdsiddique) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:317]: lets have a separate function for this logic

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/powerbi_api.py
Comment (mohdsiddique) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/powerbi_api.py:191]: why this if ?if get_modified_workspaces returns empty list then we don't need this

File: metadata-ingestion/tests/integration/powerbi/golden_test_admin_only.json
Comment (mohdsiddique) [inline metadata-ingestion/tests/integration/powerbi/golden_test_admin_only.json:2]: Could you please check why user related MCP got added, ideally these should not get ingested.

custom-properties related changes in golden file is expected.

File: metadata-ingestion/tests/integration/powerbi/golden_test_container.json
Comment (mohdsiddique) [inline metadata-ingestion/tests/integration/powerbi/golden_test_container.json:68]: same here, user should not get added in existing golden files, looks like your owner related changes are breaking previous configuration

**Review by mohdsiddique — CHANGES_REQUESTED** [review]

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:1174]: This breaks backwards compatibility for all users of powerbi stateful ingestion

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi-lexical-grammar.rule
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi-lexical-grammar.rule:438]: What are these changes for?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py
Comment (aezomz) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:1174]: Ok i will make some change so there is backward compatibility. Will need help to review the logic.

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi-lexical-grammar.rule
Comment (aezomz) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi-lexical-grammar.rule:438]: `'let\n    Source = Databricks.Catalogs("adb-123.azuredatabricks.net", "/sql/1.0/endpoints/12345dc91aa25844", [Catalog=null, Database=null]),\n    hive_metastore_Database = Source{[Name="hive_metastore",Kind="Database"]}[Data],\n    sandbox_revenue_Schema = hive_metastore_Database{[Name="sandbox_revenue",Kind="Schema"]}[Data],\n    public_consumer_price_index_Table = sandbox_revenue_Schema{[Name="public_consumer_price_index",Kind="Table"]}[Data],\n    #"Renamed Columns" = Table.RenameColumns(public_consumer_price_index_Table,{{"Country", "country"}, {"Metric", "metric"}}),\n #"Inserted Year" = Table.AddColumn(#"Renamed Columns", "ID", each Date.Year([date_id]) + Date.Month([date_id]), Text.Type),\n #"Added Custom" = Table.AddColumn(#"Inserted Year", "Custom", each Text.Combine({Number.ToText(Date.Year([date_id])), Number.ToText(Date.Month([date_id])), [country]})),\n    #"Removed Columns" = Table.RemoveColumns(#"Added Custom",{"ID"}),\n    #"Renamed Columns1" = Table.RenameColumns(#"Removed Columns",{{"Custom", "ID"}}),\n #"Filtered Rows" = Table.SelectRows(#"Renamed Columns1", each ([metric] = "Consumer Price Index") and (not Number.IsNaN([value])))\nin\n    #"Filtered Rows"',`

The for each part in this query. U can check the slack convo with mohdsiddique  here. 
https://datahubspace.slack.com/archives/C027JMRUSQJ/p1678096601528009?thread_ts=1673841991.937759&cid=C027JMRUSQJ

**Comment by aezomz** [comment]
@hsheth2  and @mohdsiddique  please help to review again. thanks!

**Comment by aezomz** [comment]
Hi @hsheth2 , can we get this reviewed again? thanks

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:346]: Please create a function `extract_dataset_schema` similar to `self.extract_lineage` and move schema generation logic into  `extract_dataset_schema`

File: metadata-ingestion/tests/integration/powerbi/golden_test_admin_access_not_allowed.json
Comment (mohdsiddique) [inline metadata-ingestion/tests/integration/powerbi/golden_test_admin_access_not_allowed.json:2]: Could you please make sure existing golden file should not get updated because of your code changes ? description related changes or new PowerBI dataset as container are acceptable

**Review by mohdsiddique — CHANGES_REQUESTED** [review]
Could you please make sure existing golden file should not get updated because of new code changes ? description related changes or new PowerBI dataset as container are acceptable

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py
Comment (aezomz) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py:509]: This function wrap the ids in a list eventually and send to PBI eitherway.
So it doesn't matter if u pass in 1 workspace_id or a List[workspace_id], in the end it will still wrap in a List and send.
Actually i recommend sending in a list of workspace to the scan api. Since PBI accept a list, they do it quite quickly too.

But if u want to think async. 
Lets make batch_size configurable then. 

https://github.com/aezomz/datahub/blob/powerbi-improvements/metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py#L1185

File: metadata-ingestion/tests/integration/powerbi/golden_test_admin_access_not_allowed.json
Comment (aezomz) [reply metadata-ingestion/tests/integration/powerbi/golden_test_admin_access_not_allowed.json:2]: Yeap! checking through. actually corpuserkey is just having inconsistent json ordering. 

I have ensured that the new feature will have no impact on existing user as much as possible.

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py
Comment (aezomz) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:1206]: > Is there any reason to skip auto_stale_entity_removal for workspaces return in modified since flow ?

modified_since will only work in this way. where we create checkpoint state for each workspace id

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py
Comment (aezomz) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:1218]: https://github.com/datahub-project/datahub/pull/7519/files/632115923077cc73b3900a1f1f3bd4c3f023bd89#r1150340668 replied here

File: metadata-ingestion/tests/integration/powerbi/golden_test_admin_access_not_allowed.json
Comment (aezomz) [reply metadata-ingestion/tests/integration/powerbi/golden_test_admin_access_not_allowed.json]: description and custom properties update are new. So are subtypes. The rest remain the same. U can open in full view and check.

**Comment by aezomz** [comment]
@hsheth2  and @mohdsiddique please help to review again. I hope we can close this soon as the default recipe setting is already backward compatible.

**Comment by aezomz** [comment]
@hsheth2  can u help to review again? we should close this soon. Can u help to resolve the conflict if all is good?

**Review by hsheth2 — COMMENTED** [review]
Have a few more comments about documentation and comments

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/config.py
Comment (aezomz) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/config.py:332]: I just checked, looks like its a radio button, only 1 can be selected. 
https://learn.microsoft.com/en-us/power-bi/collaborate-share/service-endorse-content#certify-content

File: metadata-ingestion/tests/integration/powerbi/test_m_parser.py
Comment (aezomz) [reply metadata-ingestion/tests/integration/powerbi/test_m_parser.py:35]: Yea, probably a mistake i made during rebase

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py
Comment (aezomz) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py:538]: Yup ur tweak fixed my issue as well for the Grammar, i have removed my changes.

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py
Comment (aezomz) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:255]: Fixed, i typed the data class

**Comment by aezomz** [comment]
> @aezomz in the future, it'd be much easier to review these changes if they were broken up into multiple smaller PRs

Sorry for this big PR. Thank you and Mohd so much for reviewing this as well. 

In the future, I will separate them into smaller PR so it will be more manageable. 
I have made the necessary changes, please review again. Thanks @hsheth2 !

**Comment by aezomz** [comment]
@hsheth2 please help to review again, thanks.

**Review by hsheth2 — APPROVED** [review]
LGTM

I'd like @mohdsiddique to take a final look at this before we merge it

**Comment by aezomz** [comment]
Can we merge already? @hsheth2  @mohdsiddique .

File: metadata-ingestion/tests/integration/powerbi/golden_test_container.json
Comment (hsheth2) [reply metadata-ingestion/tests/integration/powerbi/golden_test_container.json:68]: @aezomz looks like this one is still unresolved - I think this is the last pending item on this one.

File: metadata-ingestion/tests/integration/powerbi/golden_test_container.json
Comment (aezomz) [inline metadata-ingestion/tests/integration/powerbi/golden_test_container.json:502]: the corp user key is here. same as before.

File: metadata-ingestion/tests/integration/powerbi/golden_test_container.json
Comment (aezomz) [reply metadata-ingestion/tests/integration/powerbi/golden_test_container.json:68]: Same thing, i checked. the json is being re-ordered. I am not sure why. 

https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/tests/integration/powerbi/golden_test_container.json#L502

File: metadata-ingestion/tests/integration/powerbi/golden_test_container.json
Comment (aezomz) [reply metadata-ingestion/tests/integration/powerbi/golden_test_container.json:502]: @hsheth2  this is for https://github.com/datahub-project/datahub/pull/7519/files/828fb8dd1dedd243957b6a9bd5aaf291a7bc1da9#r1163567554

There are no breaking changes.

**Comment by aezomz** [comment]
Hihi @hsheth2 , lets merge this?

**Review by mohdsiddique — CHANGES_REQUESTED** [review]
Could you please address following items 

- We can remove expression from custom properties and add it in `viewProperties`, please refer https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/src/datahub/ingestion/source/looker/lookml_source.py#L935. 
The DataHub display the code in view definition section on Portal. (https://demo.datahubproject.io/dataset/urn:li:dataset:(urn:li:dataPlatform:looker,long_tail_companions.view.customer_focused,PROD)/View%20Definition?is_lineage_mode=false)
- Could you please confirm your user related changes are backward compatible ?

**Comment by mohdsiddique** [comment]
> @mohdsiddique
> 
> > We can remove expression from custom properties and add it in viewProperties, please refer https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/src/datahub/ingestion/source/looker/lookml_source.py#L935.
> > The DataHub display the code in view definition section on Portal. (https://demo.datahubproject.io/dataset/urn:li:dataset:(urn:li:dataPlatform:looker,long_tail_companions.view.customer_focused,PROD)/View%20Definition?is_lineage_mode=false)
> > Could you please confirm your user related changes are backward compatible ?
> 
> I have added view properties, and view subtype to enable the view definition.
> 
> I have fixed the user mcps ordering issues. The golden json ordering will be similar to the OSS now for corpuser.
> 
> @hsheth2 please review and lets close this. thanks

Hi @aezomz 
Could you please update the concept mapping https://github.com/aezomz/datahub/blob/powerbi-improvements/metadata-ingestion/docs/sources/powerbi/powerbi_pre.md#concept-mapping,

**Comment by aezomz** [comment]
Hi @mohdsiddique , i think its already done in this PR which is approved by Joyce. 
https://github.com/datahub-project/datahub/pull/7835

**Comment by aezomz** [comment]
Once again, please close this soon. thanks

**Comment by hsheth2** [comment]
@aezomz I'm going to merge this in tonight, assuming CI turns green

One thing I'd like to call out - the changes to stateful ingestion in this PR are definitely experimental, and I can't guarantee that we won't be making changes to stateful ingestion in the future that may mess with the implementation here.

**Pull request merged**
